### ISSUE TITLE
Internal Windows APIs are not available for UWP apps

### ISSUE THREAD

**Issue by bdbai** [tokio-rs/mio#1484]
#1034 introduced some internal Windows APIs unavailable on Universal Windows Platform, specifically `NtDeviceIoControlFile`, `RtlNtStatusToDosError`, `NtCancelIoFileEx` and `NtCreateFile`. Apps or libraries that target `*-uwp-windows-msvc/gnu` and depend on `mio` fail to link because of LNK2019, a.k.a `unresolved external symbol` of these functions. Is it possible to solve this problem without downgrading any dependencies?

**Comment by Thomasdezeeuw** [comment]
This is a duplicate of #912. For that issue there was little to no interest in supporting UWP, not sure if anything has changed.

/cc @carllerche

**Comment by bdbai** [comment]
@Thomasdezeeuw Somehow #912 was closed already. Shall we leave at least one of them open since it is really a blocking problem?

Currently we are working on a proxy app on top of UWP VPN Platform. Unlike Servo where UWP is not their main target, our implementation hugely relies on UWP. I understand that few people are interested in supporting UWP, but it is also frustrating to see the whole project blocked by such problem, given mio 0.6 still compiles for UWP targets.

May I know if you guys will accept an implementation that brings the old IOCP routines back as a fallback, enabled by certain targets or features?

**Comment by Thomasdezeeuw** [comment]
> @Thomasdezeeuw Somehow #912 was closed already. Shall we leave at least one of them open since it is really a blocking problem?

I'm ok with keeping this open, that why I didn't close it.


> Currently we are working on a proxy app on top of UWP VPN Platform. Unlike Servo where UWP is not their main target, our implementation hugely relies on UWP. I understand that few people are interested in supporting UWP, but it is also frustrating to see the whole project blocked by such problem, given mio 0.6 still compiles for UWP targets.
> 
> May I know if you guys will accept an implementation that brings the old IOCP routines back as a fallback, enabled by certain targets or features?

I would be OK with adding support for UWP, but I have two requirements:

* A person, preferably multiple, who are willing to maintain the port.
* The added code isn't too large/complex. The Windows code is complex enough as it is, adding to it's complexity is something we should avoid.

**Comment by bdbai** [comment]
Actually UWP does not need special support, as long as there is a general fallback solution when AFD is not available. Apart from UWP, This must benefit more use cases as well.

**Comment by bdbai** [comment]
For those who are interested, [here](https://github.com/YtFlow/mio-noafd) is a modified version of mio that just works on UWP. General `IoSource` and named pipes are not implemented.

**Comment by Thomasdezeeuw** [comment]
> For those who are interested, [here](https://github.com/YtFlow/mio-noafd) is a modified version of mio that just works on UWP. General `IoSource` and named pipes are not implemented.

Would you willing to maybe explain the change you've made, possibly in a longer term effort to port the changes back to Mio?

**Comment by bdbai** [comment]
> Would you willing to maybe explain the change you've made, possibly in a longer term effort to port the changes back to Mio?

Basically everything introduced in #1034 was reverted, plus some minor API changes.

I guess it would be difficult to port this back because currently Mio has a unified abstraction across different platforms. Adopting old routines means separating Windows-specific implementations from the others.

**Comment by Thomasdezeeuw** [comment]
We don't support UWP. If there is any interested in supporting UWP let us know. For now I'm closing this issue.
